### PR TITLE
OpenStack Provider: Use Neutron networking instead of Nova for Floati…

### PR DESCRIPTION
**Pull request by hmlanigan**
…ngIPs, SecurityGroups, SecurityGroupRules and Networks

### CODE REVIEW

**Comment by hmlanigan** [comment]
It was agreed by management that we would not be backwards compatible with nova networking after consultation with Canonical OpenStack folks.  The risk was deemed to be small based on juju 2.0 support statements.

We’ve also been testing against Rackspace, and other changes (api version discovery) were made to ensure that Rackspace continued to work with this change.  juju was deployed with Rackspace during the development process to ensure this.

The latest code was not tested against Rackspace as I ran into a bug where the wrong flavor was chosen so the controller instance did not deploy.  This happened in the develop branch and is unrelated to the code changes as far as i know.

Discussion of an upgrade path resulted in the consensus that it will not cause a problem, though it should be tested for the reasons you list.

The external-network config is an optional flag to allow users to choose which external network their floating ip is assigned from.  Neutron networking requires an external network be specified to create an floating ip where Nova networking did not.  Given that multiple external networks could be available, we wanted to give users a choice on which one should be used.  As well as respond to this bug: Bug 1534923: Juju needs to be able to specify floating ip pool<https://bugs.launchpad.net/juju/+bug/1534923>

If no external network is specified, the code looks for those available and in the same AZ as the instance requiring the floating ip.  From there, it choose the first one if there are multiple.  However there is still no guarantee we have chosen the best external network for the config.

On Nov 11, 2016, at 10:02 AM, Martin Packman <notifications@github.com<mailto:notifications@github.com>> wrote:


@bz2 commented on this pull request.

A couple of larger questions to resolve here before diving into the details.

First up, is just switching juju master to always using neutron what we want? This is dropping support for old/weird openstack deployments, and I'm unclear on the upgrade implications.

If I have an env I deployed with juju 2.0 and nova networking, and upgrade to 2.1, will juju have any issues from changing how it's referring to groups and floating ips? A neutron openstack should be forwarding the nova calls through anyway, but I'm a little concerned about odd behaviours here.

The rackspace provider is also built on top of this openstack code, so we'll need to make sure this doesn't break deployments against their public cloud.

Final point is I don't like the new external-network config. The current 'network' option is a hack to make things work, awaiting proper abstractions for these concepts in the juju model itself. So, we may well be stuck adding the new key for now, but it's not something we should have long term.

I will go bug some other people in the know on openstack plans on these questions.

File: provider/openstack/firewaller.go [hunk 1]
```diff
@@ -108,100 +108,130 @@ func (c *defaultFirewaller) SetUpGroups(controllerUUID, machineId string, apiPor
...
-			// are meant to apply only to this group. If you don't
-			// supply CIDR or GroupId then openstack assumes you
-			// mean CIDR=0.0.0.0/0
-			rule.GroupId = &group.Id
+	return *newGroup, nil
+}
+
+// verifyGroupRules verifies the group rules against the rules we're looking for.
+func (c *defaultFirewaller) verifyGroupRules(rules []neutron.SecurityGroupRuleV2, rulesToMatch []neutron.RuleInfoV2) bool {
+	if len(rules) - len(rulesToMatch) != 2 {
```
Comment (bz2) [inline provider/openstack/firewaller.go]: I wonder if this check is a little fragile to behaviour changes in openstack, but worst case we create another group so whatever.

File: provider/openstack/firewaller.go [hunk 2]
```diff
@@ -108,100 +108,130 @@ func (c *defaultFirewaller) SetUpGroups(controllerUUID, machineId string, apiPor
...
+		return false
+	}
+	count := len(rulesToMatch)
+	for _, rule := range rules {
```
Comment (bz2) [inline provider/openstack/firewaller.go]: Quadratic behaviour here, but we should have a pretty bounded number of rulesToMatch.

File: provider/openstack/local_test.go [hunk 3]
```diff
@@ -105,19 +107,18 @@ func registerLocalTests() {
 type localServer struct {
 	Openstack       *openstackservice.Openstack
 	Nova            *novaservice.Nova
+	Neutron         *neutronservice.Neutron
 	restoreTimeouts func()
 	UseTLS          bool
 }
 
-type newOpenstackFunc func(*identity.Credentials, identity.AuthMode, bool) (*openstackservice.Openstack, []string)
+type newOpenstackFunc func(*identity.Credentials, identity.AuthMode, bool) (*novaservice.Nova, *neutronservice.Neutron, []string)
```
Comment (bz2) [inline provider/openstack/local_test.go]: Given the args a chaning here, does the name still make sense?

File: provider/openstack/local_test.go [hunk 4]
```diff
@@ -105,19 +107,18 @@ func registerLocalTests() {
 type localServer struct {
 	Openstack       *openstackservice.Openstack
 	Nova            *novaservice.Nova
+	Neutron         *neutronservice.Neutron
```
Comment (bz2) [inline provider/openstack/local_test.go]: Do we want all three of overall openstack, nova, and neutron?

**Comment by anastasiamac** [comment]
!!build!!

**Comment by bz2** [comment]
This won't pass build check till dependencies.tsv is updated with the now-landed goose changes.

File: provider/openstack/firewaller.go [hunk 5]
```diff
@@ -108,100 +108,130 @@ func (c *defaultFirewaller) SetUpGroups(controllerUUID, machineId string, apiPor
...
+	groupsFound, err := neutronClient.SecurityGroupByNameV2(name)
 	if err == nil {
-		// Group exists, so assume it is correctly set up and return it.
-		// TODO(jam): 2013-09-18 http://pad.lv/121795
-		// We really should verify the group is set up correctly,
-		// because deleting and re-creating environments can get us bad
-		// groups (especially if they were set up under Python)
-		return *group, nil
+		for _, group := range groupsFound {
+			if c.verifyGroupRules(group.Rules, rules) {
```
Comment (hmlanigan) [inline provider/openstack/firewaller.go]: Yes,  with the caveat that the default neutron rules will also be defined.

File: provider/openstack/local_test.go (hunk 4 above)
Comment (hmlanigan) [inline provider/openstack/local_test.go]: Nova and Neutron are necessary.  Openstack is needed for control of the http services for the test double.

**Comment by bz2** [comment]
!!check!!

**Comment by bz2** [comment]
$$merge$$

**Pull request merged**